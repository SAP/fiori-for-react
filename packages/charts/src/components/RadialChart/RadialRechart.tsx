import * as ThemingParameters from '@ui5/webcomponents-react-base/lib/sap_fiori_3';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { useInitialize } from '@ui5/webcomponents-react-charts/lib/initialize';
import { ChartContainer } from '@ui5/webcomponents-react-charts/lib/next/ChartContainer';
import { PieChartPlaceholder } from '@ui5/webcomponents-react-charts/lib/PieChartPlaceholder';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import React, { CSSProperties, FC, forwardRef, Ref, useMemo } from 'react';
import { PolarAngleAxis, RadialBar, RadialBarChart as RadialBarChartLib } from 'recharts';

interface RadialChartProps extends CommonProps {
  value: number;
  maxValue?: number;
  displayValue: number | string;
  color?: CSSProperties['color'];
  height?: number | string;
  width?: number | string;
}

/**
 * <code>import { RadialChart } from '@ui5/webcomponents-react-charts/lib/next/RadialChart';</code>
 */
const RadialChart: FC<RadialChartProps> = forwardRef((props: RadialChartProps, ref: Ref<any>) => {
  const {
    maxValue = 100,
    value,
    displayValue,
    color,
    width = 300,
    height = 300,
    style,
    className,
    tooltip,
    slot
  } = props;

  useInitialize();

  const chartRef = useConsolidatedRef<any>(ref);

  const range = useMemo(() => {
    return [0, maxValue];
  }, [maxValue]);

  const dataset = useMemo(() => [{ value }], [value]);

  return (
    <ChartContainer
      dataset={dataset}
      ref={chartRef}
      placeholder={PieChartPlaceholder}
      width={width as any}
      height={height}
      style={style}
      className={className}
      tooltip={tooltip}
      slot={slot}
    >
      <RadialBarChartLib
        margin={{ left: 20, right: 20, top: 20, bottom: 20 }}
        innerRadius="90%"
        outerRadius="100%"
        barSize={10}
        data={dataset}
        cx="50%"
        cy="50%"
        startAngle={90}
        endAngle={-270}
      >
        <PolarAngleAxis type="number" domain={range} tick={false} />
        <RadialBar
          background={{ fill: ThemingParameters.sapContent_ImagePlaceholderBackground }}
          dataKey="value"
          cornerRadius="50%"
          fill={color ?? `var(--sapUiChartAccent${(0 % 12) + 1})`}
        />
        <text
          x="50%"
          y="50%"
          textAnchor="middle"
          dominantBaseline="middle"
          className="progress-label"
          style={{ fontSize: ThemingParameters.sapMFontHeader3Size, fill: ThemingParameters.sapTextColor }}
        >
          {displayValue}
        </text>
      </RadialBarChartLib>
    </ChartContainer>
  );
});

RadialChart.displayName = 'RadialChart';

export { RadialChart };
