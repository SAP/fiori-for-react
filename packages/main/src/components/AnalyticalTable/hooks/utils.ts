export const makeTemplateColumns = (columns, resizedColumns) =>
  columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
