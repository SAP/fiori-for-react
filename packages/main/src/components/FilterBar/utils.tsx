import { cloneElement, ReactElement } from 'react';

export const setPropsOfChildren = (children, currentRef) => {
  return children.map((child) => {
    if (child.props.children.props[currentRef]?.children) {
      const childrenOfChild = child.props.children;
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SELECT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-MULTI-COMBOBOX'
      ) {
        const selectedIndices = Array.from(child.props.children.props[currentRef].children)
          .map((item, index) => (item._state.selected ? index : false))
          .filter((el) => el !== false);
        const options = child.props.children.props.children.map((item, index) => {
          if (selectedIndices.includes(index)) {
            return cloneElement(item, { selected: true });
          }
          return cloneElement(item, { selected: false });
        });
        return cloneElement(child, {
          children: { ...childrenOfChild, props: { ...childrenOfChild.props, children: options } }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-INPUT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-DATEPICKER'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, value: child.props.children.props[currentRef]._state.value }
          }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SWITCH' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-CHECKBOX'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, checked: child.props.children.props[currentRef]._state.checked }
          }
        });
      }
      return child;
    }
    return child;
  });
};

export const addRef = (children, refsArray, currentRef) =>
  children.map((child) => {
    const childRef = refsArray.filter((item) => item.key === child.key);
    const childrenAttributes = child.props.children;
    const childrenProps = child.props.children.props;
    return cloneElement(child as ReactElement<any>, {
      children: { ...childrenAttributes, props: { ...childrenProps, [currentRef]: childRef[0]?.node } }
    });
  });

export const renderSearchWithValue = (searchElement, searchValue) => {
  return cloneElement(searchElement(), { value: searchValue });
};
