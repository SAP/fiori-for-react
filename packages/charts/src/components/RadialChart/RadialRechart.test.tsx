import { mountThemedComponent, renderThemedComponent } from '@shared/tests/utils';
import * as React from 'react';
import { dataset, label } from '../../resources/RechartProps';
import { RadialChart } from './RadialRechart';

describe('RadialChart', () => {
  test('Renders with data', () => {
    renderThemedComponent(<RadialChart width={'100%'} dataset={dataset} labelKey={label} />);
  });

  test('loading placeholder', () => {
    const wrapper = mountThemedComponent(<RadialChart width={'50%'} />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
