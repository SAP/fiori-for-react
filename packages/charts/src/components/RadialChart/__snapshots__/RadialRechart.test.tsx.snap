// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadialChart loading placeholder 1`] = `
<div
  style="width: 50%; height: 300px;"
>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 100%;"
  >
    <div
      style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
    />
  </div>
</div>
`;
